### PR TITLE
Feature 81: api support for updating single weekdays

### LINKED ISSUE #81: Updating order of activities when reordering in the weekplan
Related to aau-giraf/web-api#19

### PR DESCRIPTION
**Pull request by AdrianPlesner**
Closes #81 
Missing tests

### CODE REVIEW

**Comment by Determinatoro** [comment]
Please use the following checklist, @13uller. Feel free to check it off as you go along.

**Code Design** 

- [ ] The code is in the right place? (Both in terms of folder structure and class structure)
- [ ] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems     
- [ ] This code could not, to the best of my knowledge, have reused existing code
- [ ] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability** 

- [ ] Names are meaningful and self-documenting
- [ ] It is understandable, by reading the code, what it does
- [ ] The code is simple and readable. i.e. it contains no \"hack\", or obscure solution   

**Code Maintainability**

- [ ] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need to be covered by tests
- [ ] The functionality needs to be covered by tests

Only check these if the functionality needs to be covered by tests:

- [ ] All functionality is covered by tests
- [ ] The tests has sensible names
- [ ] By reading the tests you know what they cover
- [ ] The tests cover sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Comment by Determinatoro** [comment]
Please use the following checklist, @Riasalit. Feel free to check it off as you go along.

**Code Design** 

- [ ] The code is in the right place? (Both in terms of folder structure and class structure)
- [ ] The code is not over-engineered. Examples of over-engineering: Implemented behavior for future problems     
- [ ] This code could not, to the best of my knowledge, have reused existing code
- [ ] The code does not introduce functionality that is unnecessary for solving the problem

**Code Readability** 

- [ ] Names are meaningful and self-documenting
- [ ] It is understandable, by reading the code, what it does
- [ ] The code is simple and readable. i.e. it contains no \"hack\", or obscure solution   

**Code Maintainability**

- [ ] Has all the added code been properly commented

Only check one of the items below:

- [ ] The functionality does **not** need to be covered by tests
- [ ] The functionality needs to be covered by tests

Only check these if the functionality needs to be covered by tests:

- [ ] All functionality is covered by tests
- [ ] The tests has sensible names
- [ ] By reading the tests you know what they cover
- [ ] The tests cover sensible cases. Both happy paths and exception paths
- [ ] The tests cover the full functionality. i.e., The tests fails if some of the requested functionality is missing

**Comment by twaxlol** [comment]
This issue is related to issue aau-giraf/web-api#19 and issue aau-giraf/weekplanner#587

**Review by smh87 — CHANGES_REQUESTED** [review]
Needs some unit test.

**Comment by AdrianPlesner** [comment]
> Needs some unit test.

@smh87 Which tests are you suggesting that are needed? Both new methods have a test and codeCov are reporting 100%

**Review by smh87 — DISMISSED** [review]

File: lib/api/week_api.dart [hunk 1]
```diff
@@ -36,6 +38,18 @@ class WeekApi {
     });
   }
 
+  /// Get a single weekday from the specified week
+  ///
+  /// [id] User ID
+  /// [year] The year of the week
+  /// [weekNumber] The number of the week
+  /// [day] The day of the week
+  Stream<WeekdayModel> getDay(String id, int year, int weekNumber, Weekday day){
```
Comment (JeppeKLau) [inline lib/api/week_api.dart:47]: Hvorfor returnere en `Stream<T>` i stedet for et `http.Response`?

**Review by JeppeKLau — DISMISSED** [review]

File: lib/api/week_api.dart (hunk 1 above)
Comment (twaxlol) [reply lib/api/week_api.dart:47]: Fordi at api_client netop oversætter et http response til en stream. Sådan bliver det gjort i alle methods i api_client, hvorfor stiller du spørgsmålstegn ved det her?

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Fordi det virker meget forkert at styre hvad man har adgang til når man laver en http request. Det smarte her vil være at sende et http response tilbage og håndtere det i den specifikke implementation hvor man bruger requesten. Så tillader man fx at kunne reagere forskelligt på forskellige response codes.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Vi har bare fulgt standarden for projektet. Det virker lidt til at du har en idé til en refactor at hvordan api_client virker, du skal være velkommen til at lave et  nyt issue til det, men det er uden for scoret af denne pull request.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Det er bestemt indenfor scopet af den her pull request, da implementationen af featuren har krævet at metoden er blevet tilføjet og derved bliver det bestemt på det her pull request hvad metoden skal returnere

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Der virker bare forkert at implementere disse metoder helt anderledes end alle andre metoder i projektet.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Desuden er der allerede implementeret sådan at hvis svaret ikke er et SuccessRespose, bliver der smidt en ApiException der blandt andet indeholder en relevant errorcode. Alt man skal gøre er at tilføje en .onError hvor man lytter til en Stream. Dette foregår i den underliggende klasse HttpClient. (api_client/http/http_client.dart).
Så jeg kan faktisk ikke se hvad det er du synes der mangler.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Lige til den første kommentar der, så kan det være fuldstændig ligegyldigt hvordan det er implementeret i resten af clienten hvis det ikke er implementeret på en god måde.

Til det andet, så vil jeg mene at det også er en rigtig rigtig dum måde at håndtere ting på. Det at smidde en generel exception som får tekst stringe der beskriver hvad problemet er, er kode som er meget svær at bygge videre på. Det introducere problemstillingen at hvis man ændre i web apien på hvilken error besked man får tilbage. Så skal resten af koden omkring det ændres da man kigger efter en specifik fejl besked.
Modsat dette, hvis man vælger at give en mere generel svar tilbage, som fx en `http.Response`, har de fremtidige programmører på projektet mulighed for at vælge hvordan de vil bruge hele responsen og ikke bare den data der kommer retur.
Endvidere, den implementation af "error handling" du peger på sortere alle responses fra der har en response code som er højere end 300, hvilket kan være mange forskellige. Jo det bliver givet en error key tilbage, men her er der tale om noget hvor tidligere har man tænkt at man ville lave en custom implementation af noget som Flutter i forvejen kan gøre selv, se fx også `http` klassen i api clienten. Fra mit synspunkt giver disse implemtationer ingen mening da der allerede findes udemærket værktøjer til at håndtere sådanne ting i Flutter.

Men hvis du absolut vil være så stålfast på at gøre ting på en ikke standard måde og fucke med de senere studerende, skal du være min gæst. Jeg synes bare ikke det er særligt sjovt, nu hvor jeg har prøvet at være en af dem.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Hvor vidt det er implementeret på en god måde, er jo sådan set din holdning. Målet er her at implementere nye ting på en måde der passer sammen med den eksisterende kode. Hvis du mener at der er en bedre måde at gøre det på, på et generelt niveau, kan du lave et issue om at refactor hele projektet.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Det er i og for sig også min intention. Men sådan burde det også lyde for weekplanneren, dog tror jeg at jeg får lidt svære med at få medhold i den.

Men det er implementeret på en dårlig måde, for hvis du slår løsninger op på nettet der involdvere at kommunikere med internettet, i.e http requests, får du tilbage at de bruger de indbyggede klasser og ikke de her spaghetti klasser der findes i Giraf. Men lad os nu lade det ligge ved det.

Har I tænkt jer at ændre det eller lader I det ligge?

Comment (Determinatoro) [reply lib/api/week_api.dart:47]: Hej,

Jeg giver lige mit input, så den her diskussion kan blive løst.

@JeppeKLau jeg tænker, at du skal åbne et nyt issue, som beskriver den ændring, du gerne vil have. Det vil nemlig kræve en stor refactor. 

Hvis der så ikke mere, du vil have ændret, så kan du godkende den her pull request.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Vi har ingen intention om at ændre det. Det du beder om er en komplet refactor, hvilket ikke er det dette review går du på, og det er bestemt ikke noget nogen af os har tid til på dette punkt i projektet. 

Vi kan sagtes blive enige om at hele GIRAF er en stor gang spaghetti, men det er der altså ikke så meget at gøre ved nu, og det bliver bestemt ikke mindre spaghetti af at implementere denne del på en helt anden måde, om den så er bedre eller ej.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: @Determinatoro jeg skal nok åbne et nyt issue hvor jeg beskriver det I, bare rolig. Der er dog stadig noget jeg gerne vil have ændret.

@AdrianPlesner , jeg beder bestemt ikke om en komplet refactor. Det jeg beder om er at det starter med de par metoder I har implementeret her og at de bliver ændret til at give et `http.Response` tilbage. Så kan man ved der hvor metoderne bliver anvendt tage den fulde funktionalitet af Flutters indbyggede http response i brug. Og det er vigtigt at den giver den Flutter indbyggede type tilbage og ikke den fra api clienten.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Men det giver ikke mening at implementere denne del af koden helt anderledes, om det så er en bedre måde eller ej, det vil kun introducere mere spaghetti og gøre projektet endnu sværere at forstå for fremtidige studerende. Når man vil lave en så radikal ændring som du foreslår bør det være alt eller intet, hvorfor man laver en issue om at refactor.

Comment (Rosmarium) [reply lib/api/week_api.dart:47]: Det er ikke dit job som reviewer at sige, hvordan du ville have løst et problem. Det er dit job at sige, løser det problemet, er det dækket af tests og er koden læselig. Vi er faktisk utroligt ligeglade med, hvad du synes er god kode og hvordan "du synes det burde være" Hvis det løser problemet, er dækket af tests og ikke ødelægger eksisterende kode, så er alle dine meninger faktisk ligemeget.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Hvorfor er det helt præcist at det ikke giver mening at implementere det her på en anden måde end resten? Jeg kan ikke se problemet da resten af koden er meget specifik og unik, dette er det bare på sin egen måde.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Problemet er at strukturen i HELE api_client er at returnere en Stream<modelT> igennem en http_Client der smider ApiException hvis det går galt, og det er altid bedst at følge strukturen af den eksisterende kode, om den så er god eller ej. Hvis strukturen skal ændres, gør man det på hele projektet. Jeg har ikke tænkt mig at bryde strukturen i mine metoder fordi du synes det er en dårlig struktur.

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Det er også helt fair at du ikke har lyst til at ændre din løsning. Men som @Rosmarium så fint udpegede, er "[...] meninger faktisk ligemeget." Endvidere, hvis du ikke ændre din løsning, kan jeg ikke checke af i alle kasser på review guiden og derfor bør denne løsning ikke pulles ind i develop.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: Hvilket punkt er det helt præcist du mener ikke er opfyldt, og hvad er din præcise begrundelse for det?

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: Jeg mener at punktet om obskure løsninger ikke er opfyldt, jævnfør diskussionen herover.
Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.

Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: > Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.

Det er første gang jeg hører om det?

Comment (JeppeKLau) [reply lib/api/week_api.dart:47]: > 
> 
> > Et andet punkt der heller ikke er opfyldt et punktet om at tests dækker både gode og dårlige test cases.
> 
> Det er første gang jeg hører om det?

Det er meget muligt, men punktet har aldrig været tjekket af på review listen.

**Comment by AdrianPlesner** [comment]
Er tests tilfredstillende nu?

**Comment by JeppeKLau** [comment]
> 
> 
> Er tests tilfredstillende nu?

Jeg er lidt i tvivl. Det virker til at du bare siger til den at den skal forvente noget på en url og siger til den at den skal kaste en fejl, uden at den implementerede kode bliver kørt. Men det er meget muligt jeg har misforstået hvordan Flutter kører sine tests.

File: lib/api/week_api.dart (hunk 1 above)
Comment (AdrianPlesner) [reply lib/api/week_api.dart:47]: > > Er tests tilfredstillende nu?
> 
> Jeg er lidt i tvivl. Det virker til at du bare siger til den at den skal forvente noget på en url og siger til den at den skal kaste en fejl, uden at den implementerede kode bliver kørt. Men det er meget muligt jeg har misforstået hvordan Flutter kører sine tests.

Jeg tror bare det er sådan det er. Vi kan jo ikke have en faktisk forbindelse til et api med data, derfor har vi en mock vi setter til at kaste fejl, og så assert at den fejl kommer.

**Comment by AdrianPlesner** [comment]
> > Er tests tilfredstillende nu?
> 
> Jeg er lidt i tvivl. Det virker til at du bare siger til den at den skal forvente noget på en url og siger til den at den skal kaste en fejl, uden at den implementerede kode bliver kørt. Men det er meget muligt jeg har misforstået hvordan Flutter kører sine tests.

Jeg tror bare det er sådan det er. Vi kan jo ikke have en faktisk forbindelse til et api med data, derfor har vi en mock vi setter til at kaste fejl, og så assert at den fejl kommer.

**Comment by JeppeKLau** [comment]
> 
> 
> > > Er tests tilfredstillende nu?
> > 
> > 
> > Jeg er lidt i tvivl. Det virker til at du bare siger til den at den skal forvente noget på en url og siger til den at den skal kaste en fejl, uden at den implementerede kode bliver kørt. Men det er meget muligt jeg har misforstået hvordan Flutter kører sine tests.
> 
> Jeg tror bare det er sådan det er. Vi kan jo ikke have en faktisk forbindelse til et api med data, derfor har vi en mock vi sætter til at kaste fejl, og så assert at den fejl kommer.

Det er rigtigt nok og der er jo noget som gerne skulle fange fejlen inden den bliver sendt gennem streamen, hvilket jeg antager er blevet testet et andet sted, så det fint.

File: lib/api/week_api.dart (hunk 1 above)
Comment (smh87) [reply lib/api/week_api.dart:47]: system-archi

Comment (smh87) [reply lib/api/week_api.dart:47]: skal vi ikke lukke den her og få en ordenligt fagligt tune her fra?

Comment (smh87) [reply lib/api/week_api.dart:47]: evt. Discord samtale eller slack chat?

**Review by smh87 — APPROVED** [review]
Screenshot from 2020-12-03 22-08-39
approved since the week_api has 100% code coverage, and is tested with the other branches too.

**Review by social2dk — APPROVED** [review]

**Comment by JeppeKLau** [comment]
@LivHolm det er dejligt at se at samarbejdet på det her projekt er faldet så meget fra hinanden at når man ønsker ændringer som vil forbedre programmet og dem der skal lave ændringerne bare sætter sig på tværs og ikke vil lave dem, så omgår vi bare dem der vil have lavet ændringerne og tvinger dårlige implementationer igennem.

**Comment by LivHolm** [comment]
> @LivHolm det er dejligt at se at samarbejdet på det her projekt er faldet så meget fra hinanden at når man ønsker ændringer som vil forbedre programmet og dem der skal lave ændringerne bare sætter sig på tværs og ikke vil lave dem, så omgår vi bare dem der vil have lavet ændringerne og tvinger dårlige implementationer igennem.

Hej @JeppeKLau der blev i fællesskab taget en beslutning til Sprint Review i forhold til hvordan alle aktive pull requests skulle håndteres. De ændringer du gerne vil have lavet må du meget gerne lave et nyt issue på.

**Pull request merged**